### PR TITLE
Matrix math optimizations for memory reuse. 

### PR DESCRIPTION
**Pull request by jasonimercer**
New matrix math implemented and used where the results are stored in a supplied matrix rather than generating a new temporary matrix for each math operation.

Provides approx 30% savings in test area use-case. 

CORE-1730

### CODE REVIEW

**Comment by jasonimercer** [comment]
@skaynama @servos

File: acado/matrix_vector/block_matrix.cpp [hunk 1]
```diff
@@ -108,6 +152,40 @@ BlockMatrix BlockMatrix::operator+( const BlockMatrix& arg ) const{
 	return tmp;
 }
 
+void BlockMatrix::mat_mat_add(const BlockMatrix& m1, const BlockMatrix& arg, BlockMatrix& result)
+{
```
Comment (jasonimercer) [inline acado/matrix_vector/block_matrix.cpp:156]: All the matrix math is a copy-paste of existing methods and tweaked for static functions and provided results matrix.

File: acado/nlp_solver/scp_evaluation.cpp [hunk 2]
```diff
@@ -476,10 +499,19 @@ double SCPevaluation::getKKTtolerance(	const OCPiterate& iter,
 
     if( ( constraint != 0 ) || ( dynamicDiscretization != 0 ) )
     {
-        ((cp.lambdaConstraint.getAbsolute()).addRegularisation(eps)^cp.upperConstraintResiduum.getNegative()).getSubBlock( 0, 0, tmp, 1, 1 );
+        cp.lambdaConstraint.getAbsolute(absLC);
+		absLC.addRegularisation(eps);
```
Comment (jasonimercer) [inline acado/nlp_solver/scp_evaluation.cpp:503]: ugly indenting - ACADO's source files use all the indenting conventions simultaneously. Will fix.

**Comment by skaynama** [comment]
LGTM, but let's do some tests first.

**Pull request merged**